[maiyasaliha] iP

maiyasaliha opened the pull request:
### Just Ken

> "What if there's beach? You'll need someone professional in that." — Ken

Your very own Long Term, Long Distance, Low Commitment, Casual Botfriend ;)

Ken frees your mind of having to remember things you need to do, so you can focus on beach🏖. 
It's,

- text-based
- easy to learn
- **SUPER EASY** to use

All you need to do is,

1. download it from here.
2. run.
3. add your tasks.
4. and _Beach Off!_

Features:

- [ ] Managing tasks
- [ ] Managing deadlines
- [ ] Reminders (coming soon) 🛼

Try It Now! - [and hear what Ken has to say!](https://youtu.be/5peWZA39yNE?si=y5PYe1uKV1kn9Syn) 💅🏽

For all Java Programmers, nothing much, just look at the main method!

```
public static void main(String[] args) {
        new Ken(FILE_PATH).run();
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

maiyasaliha committed:
Level 0: Added Hi and Bye

maiyasaliha committed:
Level 1: Echoing complete

maiyasaliha committed:
Level 2: Add, List Completed

maiyasaliha committed:
Task Class Created

maiyasaliha committed:
Level 3: Mark as Done

maiyasaliha committed:
Todo Class Created

maiyasaliha committed:
Event Class Created

maiyasaliha committed:
Deadline Class Created

maiyasaliha committed:
Todo Edited

maiyasaliha committed:
Level 4: ToDo, Event, Deadline Completed

maiyasaliha committed:
KenException Created

maiyasaliha committed:
Level 5: Handle Errors Completed

maiyasaliha committed:
Level 6: Delete Completed

maiyasaliha committed:
Automated Text UI Testing Complete

maiyasaliha committed:
Bugs in bye Fixed

maiyasaliha committed:
Update Error message in Ken.java

maiyasaliha committed:
Update Deadline, Event, Task, Todo to accomodate standardised writing and reading of task
Update Ken to Store in and Read Tasks from data/ken
Create data/ken.txt for storage

maiyasaliha committed:
Update Ken to not throw error when data/ken.txt does not exist, create directory and work nicely

maiyasaliha committed:
Merge branch 'branch-Level-7'

maiyasaliha committed:
Level 8: Update Deadline and Event to store Time in Time Format

maiyasaliha committed:
Merge branch 'branch-Level-8'

maiyasaliha committed:
Create Ui Class with user interacting messages

maiyasaliha committed:
Create TaskList Class to store tasks in list

maiyasaliha committed:
Create Storage Class to store and load tasks to and from harddrive

maiyasaliha committed:
Create Parser class to make sense of commands and get further info if needed

maiyasaliha committed:
AMoreOOP: Separate all code in Ken to respective classes and integrate them back

maiyasaliha committed:
Merge branch 'branch-AMoreOOP'

maiyasaliha committed:
A-Packages: Create packages Ken, task, ui, parser, storage

maiyasaliha committed:
Update classes to accomodate packaging

maiyasaliha committed:
Merge branch 'branch-A-Packages'

maiyasaliha committed:
Merge branch 'add-gradle-support'

maiyasaliha committed:
A-Gradle: Gradle build

maiyasaliha committed:
Refactor package

maiyasaliha committed:
respective changes made in files due to package refactoring

maiyasaliha committed:
created working jar

maiyasaliha committed:
Merge branch 'branch-A-Jar'

maiyasaliha committed:
Add JavaDocs for all classes

JavaDocs has been added for all classes and all methods

maiyasaliha committed:
Add Finding feature
Ken is now able to seek out tasks with keywords specified by user and displays them

maiyasaliha committed:
Modify classes to comply with coding standard

Modify Storage, Task and TaskList to comply with coding standard. Everthing else is already according to coding standard.

maiyasaliha committed:
Merge branch 'branch-Level-9'

maiyasaliha committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/Ken/task/Task.java
#	src/main/java/Ken/task/TaskList.java

Merge conflict has been resolved to include changes from both branches.

maiyasaliha committed:
Merge branch 'branch-A-CodingStandard'

maiyasaliha committed:
Create Test Classes

Test classes created to test methods in Ui and TaskList classes.

There are still some issues persisting:
Although expected output and actual output is exactly the same, assert equals is failing for some cases.
Have tried to overcome it by trimming white space, but have to take a look at it again.

tzyian commented on src/main/java/Ken/parser/Parser.java:
```diff
@@ -0,0 +1,88 @@
+        this.scanner = new Scanner(System.in);
+        this.taskList = taskList;
+        this.storage = storage;
+        this.saidBye = false;
```
Should this be `hasSaidBye`?

tzyian commented on src/main/java/Ken/parser/Parser.java:
```diff
@@ -0,0 +1,88 @@
+                System.out.println("where you going?\n");
+            } else if (!command.equals("bye")) {
+                System.out.println("don't know what that is\n");
+            } else {
+            }
```
Is there a need for an empty else clause?

tzyian commented on src/main/java/Ken/parser/Parser.java:
```diff
@@ -0,0 +1,88 @@
+            } else {
+            }
+
+        } while (!command.equalsIgnoreCase("bye"));
+        storage.saveTasks(taskList.getTasks());
```
Would it be better to leave a newline in between?

tzyian commented on src/main/java/Ken/task/Deadline.java:
```diff
@@ -0,0 +1,49 @@
+     */
+    @Override
+    public String toString() {
+        return "[D]" + super.toString() + " (by: " + this.by.format(DateTimeFormatter.ofPattern("MMM dd yyyy HH:mm")) + ")";
```
Would it be better if you extract these formatters to a separate method?

tzyian commented on src/main/java/Ken/task/Event.java:
```diff
@@ -0,0 +1,61 @@
+     */
+    @Override
+    public String toString() {
+        return "[E]" + super.toString() + " (from: " + this.from.format(DateTimeFormatter.ofPattern("MMM dd yyyy HH:mm")) +
```
Would it be better to split this line to make it shorter?
Looks like a recurring issue

tzyian commented on src/main/java/Ken/task/Task.java:
```diff
@@ -0,0 +1,110 @@
+     * @return The status icon ("✔" if done, " " if not done).
+     */
+    private String getStatusIcon() {
+        return (isDone ? "✔" : " "); // mark done task with X
```
Your comment seems to be different from your code?

tzyian commented on src/main/java/Ken/task/TaskList.java:
```diff
@@ -0,0 +1,223 @@
+     * @param description The description of the event task.
+     * @throws KenException If the description or event command is invalid.
+     */
+    public void addEventTask(String description) throws KenException {
+        try {
+            int indexOfFrom = description.indexOf("/from");
+            int indexOfTo = description.indexOf("/to");
+
+            if (indexOfFrom != -1 && indexOfTo != -1) {
+                String eventDescription = description.substring(0, indexOfFrom).trim();
+                String from = description.substring(indexOfFrom + 5, indexOfTo).trim();
+                String to = description.substring(indexOfTo + 3).trim();
+                Event event = new Event(eventDescription, from, to);
+                addTask(event);
```
I didn't check the logic, but is it guaranteed that there will be spaces before the `/from` and `/to`? 
Eg do you consider `todos taskName/byNOW/toNOW` a valid input?
This is up to you, just in case you didn't consider it

tzyian commented on src/main/java/Ken/task/TaskList.java:
```diff
@@ -0,0 +1,223 @@
+     *
+     * @param index The index of the task to be unmarked as done.
+     */
+    public void unmarkTask(int index) {
```
Might want to be a bit more explicit in method name of what you mean by marking and unmarking a task?

tzyian commented on src/main/java/Ken/ui/Ui.java:
```diff
@@ -0,0 +1,24 @@
+    /**
+     * Displays a goodbye message to the user.
+     */
+    public void byeMessage() {
```
Perhaps a more explicit method name? Elsewhere too

tzyian commented on src/main/java/Ken/ui/Ui.java:
```diff
@@ -0,0 +1,24 @@
+package ken.ui;
+
+/**
+ * The Ui class is responsible for handling user interface-related messages.
+ * It provides methods to display welcome and goodbye messages to the user.
+ */
+public class Ui {
```
If you already have a Ui class, would it be better to shift all `System.out::println` from other classes to here?

tzyian commented on src/main/java/Ken/parser/Parser.java:
```diff
@@ -0,0 +1,88 @@
+        do {
+            command = scanner.nextLine();
+
+            if (command.equals("list")) {
+                taskList.listTasks();
+            } else if (command.startsWith("mark ")) {
+                taskList.markTask(Integer.parseInt(command.substring(5)));
+            } else if (command.startsWith("unmark ")) {
+                taskList.unmarkTask(Integer.parseInt(command.substring(7)));
+            } else if (command.startsWith("delete ")) {
+                taskList.deleteTask(Integer.parseInt(command.substring(7)));
+            } else if (command.startsWith("todo ")) {
+                taskList.addTodoTask(command.substring(5));
+            } else if (command.startsWith("deadline ")) {
+                taskList.addDeadlineTask(command.substring(9));
+            } else if (command.startsWith("event ")) {
+                taskList.addEventTask(command.substring(6));
+            } else if (command.equalsIgnoreCase("bye")) {
+                saidBye = true;
+            } else if (command.startsWith("seek ")) {
+                taskList.findTasks(command.substring(5));
+            } else if (command.startsWith("seek")) {
+                System.out.println("OKAY I'LL GO HIDE!\n");
+            } else if (command.startsWith("todo")) {
+                System.out.println("do what?\n");
+            } else if (command.startsWith("deadline")) {
+                System.out.println("oh no! which line died?\n");
+            } else if (command.startsWith("event")) {
+                System.out.println("where you going?\n");
+            } else if (!command.equals("bye")) {
+                System.out.println("don't know what that is\n");
```
I like how your if-elses are easy to read

tzyian reviewed (changes requested):
There are a few coding standard violations. Otherwise LGTM!

NewtonKoh reviewed (commented):
LGTM! code is quite easy to follow and functions/variables are well named!

maiyasaliha committed:
Add GUI
GUI has been successfully added
Still facing some issues with last part of Level-10 fxml and separating ken class into its constituents, will come back to that later

maiyasaliha committed:
Update code to comply with Code Quality

There are several methods that print task for deadline and event that have date formatting in the method.
Extracting out date formatting in printing methods and putting them in a separate method improves code quality.

Deleting dead code also improves code quality.

As a step towards such improvement of code quality, let's extract all such methods into separate methods in their respective classes.

Doing so will improve readability.

maiyasaliha committed:
Add assertions

Using assert feature to document important assumptions that should hold at various points in the code.
Such as the by for deadline, from and to for event not being null.

maiyasaliha committed:
Merge pull request #1 from maiyasaliha/branch-A-Assertions

Add assertions

Using assert feature to document important assumptions that should hold at various points in the code. Such as the by for deadline, from and to for event not being null.

maiyasaliha committed:
Merge branch 'master' of https://github.com/maiyasaliha/ip

maiyasaliha committed:
Merge branch 'branch-A-CodeQuality'

maiyasaliha committed:
Delete Deadcode

Updated codebase to delete deadcode

maiyasaliha committed:
Add Help Feature

Added help command so that when users are unsure of what to do or how to use ken, they now have a command to use to get all details of possible commands available that they can give ken.

maiyasaliha committed:
Merge branch 'branch-A-BCDExtensions'

maiyasaliha committed:
Delete Deadcode

More deadcode deleted to imporve code quality

maiyasaliha committed:
Merge branch 'branch-A-CodeQuality'

# Conflicts:
#	src/main/java/Ken/ui/Ui.java

maiyasaliha committed:
Update to include fxml files.

Ken was running solely on Ken.java.
Now has been abstracted out to run on Main, MainWindow, DialogBox.

Add view and fxml files to contain design instead on java files.

maiyasaliha committed:
Merge branch 'branch-fxml'

maiyasaliha committed:
Remove unnecessary import statement.

maiyasaliha committed:
Add Ui.png

Screenshot of Ui has been added.

maiyasaliha committed:
Update README.md

Update README.md to have more detail about Ken

maiyasaliha committed:
Update User Guide

User Guide has been updated.
Now includes all features of Ken and how to use Ken.

maiyasaliha committed:
Merge branch 'master' of https://github.com/maiyasaliha/ip

# Conflicts:
#	docs/README.md

maiyasaliha committed:
Update User Guide

User Guide has been updated.
Now includes all commands to given ken.
And outputs from ken.

Full user guide with picture available.

maiyasaliha committed:
Update README

Updated README to have Leaving app fetaure as last.
build.gradle modified to be able to get fat jar.

TaskList modified to display additional error message.
Now asks if help is required.

maiyasaliha committed:
Update Parser

Updated parser such that program now exits with a 1 second delay.
Not as immeditae as it used to be now, easier to read bye message.

maiyasaliha committed:
Update code to follow coding standard.

Deleted some deadcode.
Added javadocs to more methods.
Updated J Unit test cases.

maiyasaliha committed:
Merge branch 'branch-A-finalChanges'

maiyasaliha committed:
Update DialogBox and README

Fixed bug in text wrapping, now user input nicely formats to next line.
Does not keep going in a straight line.

 Updated README to include fun features.

maiyasaliha committed:
Finalise features

Updated mark and unmark to not be be able to double mark or unmark.
Now they both only do it once.
And have a different message when double marked or double unmarked.

maiyasaliha committed:
Update Ui.java

Updated seek to have different response where there are no mathing items in list.
Now getDisplayMatchingTasksMessages gives different message where no tasks match.

maiyasaliha committed:
Modify data/ken.txt

Git commits are violating the standard.

Making trivial changes to get latest 5 commits to follow standard.

Commit 1.

maiyasaliha committed:
Modify data/ken.txt

Git commits are violating the standard.

Making trivial changes to get latest 5 commits to follow standard.

Commit 2.

maiyasaliha committed:
Modify data/ken.txt

Git commits are violating the standard.

Making trivial changes to get latest 5 commits to follow standard.

Commit 3.

maiyasaliha committed:
Modify data/ken.txt

Git commits are violating the standard.

Making trivial changes to get latest 5 commits to follow standard.

Commit 4.

maiyasaliha committed:
Modify data/ken.txt

Git commits are violating the standard.

Making trivial changes to get latest 5 commits to follow standard.

Commit 5.
